ledger/acctupdates_test.go

```diff
@@ -1352,6 +1352,191 @@ func TestBoxNamesByAppIDs(t *testing.T) {
 	}
 }
 
+func TestBoxCache(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	initialBlocksCount := 1
+	accts := make(map[basics.Address]basics.AccountData)
+
+	protoParams := config.Consensus[protocol.ConsensusCurrentVersion]
+	protoParams.MaxBalLookback = 8
+	protoParams.SeedLookback = 1
+	protoParams.SeedRefreshInterval = 1
+	testProtocolVersion := protocol.ConsensusVersion("test-protocol-TestBoxNamesByAppIDs")
+	config.Consensus[testProtocolVersion] = protoParams
+	defer func() {
+		delete(config.Consensus, testProtocolVersion)
+	}()
+
+	ml := makeMockLedgerForTracker(t, true, initialBlocksCount, testProtocolVersion,
+		[]map[basics.Address]basics.AccountData{accts},
+	)
+	defer ml.Close()
+
+	conf := config.GetDefaultLocal()
+	au, _ := newAcctUpdates(t, ml, conf, ".")
+	defer au.close()
+
+	knownCreatables := make(map[basics.CreatableIndex]bool)
+	opts := auNewBlockOpts{ledgercore.AccountDeltas{}, testProtocolVersion, protoParams, knownCreatables}
+
+	boxCnt := 1000
+	boxesPerBlock := 100
+	curBox := 0
+	var currentRound basics.Round
+	currentDBRound := basics.Round(1)
+
+	boxMap := make(map[string]string)
+	for i := 0; i < boxCnt; i++ {
+		boxMap[fmt.Sprintf("%d", i)] = fmt.Sprintf("value%d", i)
+	}
+
+	// add boxesPerBlock boxes on each iteration
+	for i := 0; i < boxCnt/boxesPerBlock+int(conf.MaxAcctLookback); i++ {
+		currentRound = currentRound + 1
+
+		kvMods := make(map[string]*string)
+		if i < boxCnt/boxesPerBlock {
+			for j := 0; j < boxesPerBlock; j++ {
+				name := fmt.Sprintf("%d", curBox)
+				curBox++
+				val := boxMap[name]
+				kvMods[name] = &val
+			}
+		}
+
+		auNewBlock(t, currentRound, au, accts, opts, kvMods)
+		auCommitSync(t, currentRound, au, ml)
+
+		// ensure rounds
+		rnd := au.latest()
+		require.Equal(t, currentRound, rnd)
+		if uint64(currentRound) > protoParams.MaxBalLookback {
+			require.Equal(t, basics.Round(uint64(currentRound)-protoParams.MaxBalLookback), au.cachedDBRound)
+		} else {
+			require.Equal(t, basics.Round(0), au.cachedDBRound)
+		}
+
+		// verify cache doesn't contain the new boxes until commited to DB
+		if i < boxCnt/boxesPerBlock {
```

Comment:
@algoidurovic Can you help me understand the significance of the if-statement?

When I remove these statements wrapping `kvMods` check, the test still passes.  It makes me think the if-statement is extraneous.

Reply:
I think it technically is extraneous. Boxes are only being manipulated for `boxCnt/boxesPerBlock` blocks, and the remaining iterations of the loop are meant to verify the cache is in the right state after all the blocks have been committed. I'll refactor this into two loops to make it more clear.

Reply:
Actually, I take it back. The body of the loop can't be divided up easily. I'll just add more comments to describe what's going on.

Reply:
@algoidurovic Resolving - Thanks for the inline comment + test implementation change here.